### ISSUE TITLE
Crash on OS X on premature connection close

### ISSUE THREAD

**Issue by onitake** [onitake/daliserver#12]
Previously reported by @ben-knt on issue #11.

@ben-knt
By the way, I went through the Perl scripts and have some comments. Lampoff.pl crashes daliserver... for what reason, I don't know? I'm happy to help find out why. We can discuss offline sometime if you would like.

@onitake

> Uh, what?
> I know that `usbdali.pm` is a bit buggy and doesn't return the response correctly. I fixed that problem locally, but haven't pushed it to github yet.
> 
> Are you positive it crashes daliserver?
> 
> Ok, fixed `usbdali.pm` in bae41b12.

@ben-knt

I pulled in the changed file but no change. Lampoff.pl successfully turns off a lamp but the server crashes after.  Two commands issued below:

Braque:perl benknuth$ perl allset.pl 255
Writing 4 bytes to socket, address=254 command=254
Received status:success response:0
Braque:perl benknuth$ perl lampoff.pl 12
Writing 4 bytes to socket, address=25 command=0
Braque:perl benknuth$ 

[2016-00-22 20:28:13] INFO Starting daliserver
[2016-00-22 20:28:13] INFO Server ready, waiting for events
[2016-00-22 20:28:32] INFO Got connection from 127.0.0.1:27089
[2016-00-22 20:28:32] INFO Got frame: 0x02 0x00 0xfe 0xfe
[2016-00-22 20:28:32] INFO Enqueued transfer (0x7fb7abc06cc0,0x7fb7abc06a90)
[2016-00-22 20:28:32] INFO Sending data to device:
0x00000000 12 01 00 03 00 00 fe fe 00 00 00 00 00 00 00 00 ................
0x00000010 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ................
0x00000020 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ................
0x00000030 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ................
Direction: USB<->DALI Sequence number: 01 Type: 16bit DALI Address: fe Command: fe 
[2016-00-22 20:28:32] INFO Received data from device:
0x00000000 12 73 00 00 fe fe ff ff 01 00 00 00 00 00 00 00 .s..............
0x00000010 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ................
0x00000020 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ................
0x00000030 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ................
Direction: USB<->DALI Type: Send complete Address: fe Command: fe Status: ffff 
[2016-00-22 20:28:32] INFO Received data from device:
0x00000000 12 71 00 00 00 00 00 8a 01 00 00 00 00 00 00 00 .q..............
0x00000010 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ................
0x00000020 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ................
0x00000030 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ................
Direction: USB<->DALI Type: Send response Response: ff Status: 008a Sequence number: 01 
[2016-00-22 20:28:32] INFO Response to (0x00 0x00): 0xff [0x008a]
[2016-00-22 20:28:32] INFO Connection 9 was disconnected
[2016-00-22 20:28:32] INFO Closing connection 9
[2016-00-22 20:28:45] INFO Got connection from 127.0.0.1:27345
[2016-00-22 20:28:45] INFO Got frame: 0x02 0x00 0x19 0x00
[2016-00-22 20:28:45] INFO Enqueued transfer (0x7fb7abe01520,0x7fb7abe03090)
[2016-00-22 20:28:45] INFO Connection 9 was disconnected
[2016-00-22 20:28:45] INFO Closing connection 9
[2016-00-22 20:28:45] INFO Sending data to device:
0x00000000 12 02 00 03 00 00 19 00 00 00 00 00 00 00 00 00 ................
0x00000010 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ................
0x00000020 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ................
0x00000030 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ................
Direction: USB<->DALI Sequence number: 02 Type: 16bit DALI Address: 19 Command: 00 
[2016-00-22 20:28:45] INFO Received data from device:
0x00000000 12 73 00 00 19 00 ff ff 02 00 00 00 00 00 00 00 .s..............
0x00000010 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ................
0x00000020 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ................
0x00000030 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ................
Direction: USB<->DALI Type: Send complete Address: 19 Command: 00 Status: ffff 
[2016-00-22 20:28:45] INFO Received data from device:
0x00000000 12 71 00 00 00 00 00 8a 02 00 00 00 00 00 00 00 .q..............
0x00000010 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ................
0x00000020 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ................
0x00000030 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ................
Direction: USB<->DALI Type: Send response Response: ff Status: 008a Sequence number: 02 
[2016-00-22 20:28:45] INFO Response to (0x00 0x00): 0xff [0x008a]
[2016-00-22 20:28:45] **ERROR** Error writing 4 bytes to connection 2059272945: Bad file descriptor
Segmentation fault: 11
Braque:~ benknuth$

**Comment by onitake** [comment]
Ouch!
I'll look into this right away. It's clear that the connection is killed before the response can be sent, as some of the Perl scripts don't read the response at all. But the crash is in no way ok!

Can you run daliserver in gdb and produce a backtrace?

**Comment by onitake** [comment]
By @ben-knt:

Here's the backtrace. Hope this helps! 
https://gist.github.com/anonymous/bced4b4bc594748b0d3f

These two commands sent were: 
perl lampset.pl 12 255
Writing 4 bytes to socket, address=24 command=254
Received status:success response:0
perl lampoff.pl 12
Writing 4 bytes to socket, address=25 command=0

**Comment by onitake** [comment]
Umm... I didn't get your backtrace. Did you forget the link? _Edit:_ Link fixed

But no matter, I found the bug. It's just a bit hard to fix. Need to think about it some more.

As a temporary workaround, I pushed a modified lampoff.pl that checks the server response. This should prevent the crash.

**Comment by ben-knt** [comment]
Yes, pulled in the modified script and it works fine.

**Comment by onitake** [comment]
I just pushed some changes that should ensure proper dequeuing of closed connections.

Can you test them on your system using the following Perl script?

``` perl
use usbdali;
my $host = $ARGV[0] || '127.0.0.1';
my $dali = usbdali->new($host);
$dali->connect() or die("Can't connect");
$dali->send(usbdali->make_cmd('broadcast', 'up'));
$dali->disconnect();
```

**Comment by ben-knt** [comment]
Rebuilt with the changes. Below what I'm seeing with the script

[2016-01-10 16:24:05] INFO Got connection from 127.0.0.1:52421
[2016-01-10 16:24:05] INFO Got frame: 0x02 0x00 0xff 0x08
[2016-01-10 16:24:05] INFO Enqueued transfer (0x7fb0a1500dd0,0x7fb0a1500d80)
[2016-01-10 16:24:05] INFO Connection 9 was disconnected
[2016-01-10 16:24:05] INFO Closing connection 9
[2016-01-10 16:24:05] INFO Sending data to device:
0x00000000 12 0c 00 03 00 00 ff 08 00 00 00 00 00 00 00 00 ................
0x00000010 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ................
0x00000020 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ................
0x00000030 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ................
Direction: USB<->DALI Sequence number: 0c Type: 16bit DALI Address: ff Command: 08 
[2016-01-10 16:24:05] INFO Received data from device:
0x00000000 12 73 00 00 ff 08 ff ff 0c 00 00 00 00 00 00 00 .s..............
0x00000010 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ................
0x00000020 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ................
0x00000030 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ................
Direction: USB<->DALI Type: Send complete Address: ff Command: 08 Status: ffff 
[2016-01-10 16:24:05] INFO Received data from device:
0x00000000 12 71 00 00 00 00 00 8a 0c 00 00 00 00 00 00 00 .q..............
0x00000010 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ................
0x00000020 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ................
0x00000030 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ................
Direction: USB<->DALI Type: Send response Response: ff Status: 008a Sequence number: 0c 
[2016-01-10 16:24:05] INFO Response to (0x00 0x00): 0xff [0x008a]

**Comment by onitake** [comment]
I may still have to run some tests with valgrind to check if everything is released properly, but at least the references to deallocated memory are gone. This was a pretty serious bug.

I'm going to close the issue now, please reopen if you see any related crashes, unprocessed requests or lockups.

Fixed by: 9ee4874c, 396301b4 and 95a121ff
